### DISCUSSION TITLE
Multiple effect passes issues

### DISCUSSION THREAD

**Discussion by swingingtom** [pmndrs/postprocessing#342] [Q&A]
Hey postprocessing community,
Im currently having troubles to combine multiple effect passes on the composer.

I've set 2 scenes ( background & foreground ) and I try to exclusively set one effect pass on Back, and one effect pass on Front.

But the second effect pass, seems to be altered by the first one. Even if there is a clear pass between them.
Here is a link to the trouble I have : https://codesandbox.io/s/multiple-effect-passes-issue-htut8?file=/src/App.js

By using the gui, we can see that disabling `effectPass#1`(dof+glitch) impacts `effectPass#2`(godRays)
We can also show the depth texture ( not the one used for godRays, but a reference ) and see that effectPass#1 one take over effectPass#2, and then the depth is from effectPass#1 and not effectPass#2.

I currently have 2 leads, but Im blocked to investiguate those

## Effect Execution Order
As documented [here](https://github.com/vanruesc/postprocessing/wiki/Effect-Merging#effect-execution-order), effects are sorted by their priority. But as I understand, they are sorted within an EffectPass. It this provided case, Their is two EffectPasses. Is this sort order still apply ?

Is there a way to ask the composer to show/tell the final execution order of all passes and effects added?

## Convolution + Convolution
As documented [here](https://github.com/vanruesc/postprocessing/wiki/Effect-Merging#convolution--convolution) sometime it might requires to have multiple input buffers in order to achieve the proper result.

```javascript
// Two passes (right way).
const outputBuffer = [0, 0, 0, 0];
outputBuffer[0] = blur(inputBuffer[0], 0, inputBuffer); // => 40
outputBuffer[1] = blur(inputBuffer[1], 1, inputBuffer); // => 105
outputBuffer[2] = blur(inputBuffer[2], 2, inputBuffer); // => 127.33
outputBuffer[3] = blur(inputBuffer[3], 3, inputBuffer); // => 169.66
```

>Solving more advanced compositing problems sometimes requires knowledge about the inner workings of the EffectComposer. The current pipeline is flexible enough to get most image compositions done, but it's not exactly easy to use and the details about the internal buffer management are undocumented since they are supposed to be private despite being relevant. This is a design flaw and the system is error-prone (https://github.com/vanruesc/postprocessing/issues/225).

I know @vanruesc is working on the InputBuffer management api, but in the meantime, does anyone have an example/sample to share on how manually setting inputBuffers can be perform in the current api?  

Thanks!

**Comment by vanruesc** [comment]
> But as I understand, they are sorted within an EffectPass.

That's correct.

> Is there a way to ask the composer to show/tell the final execution order of all passes and effects added?

Not yet, but I'm planning on adding a `Pipeline` wrapper with an `explain` method that will provide more detailed debug information.

> We can also show the depth texture ( not the one used for godRays, but a reference ) and see that effectPass#1 one take over effectPass#2, and then the depth is from effectPass#1 and not effectPass#2.

I think you're hitting #225. The problem is that the depth texture is only attached to the first frame buffer. This works with common setups, but breaks in weird ways with more complicated compositions. The planned buffer management changes will fix that.

Changing the `needsSwap` flag of the effect passes didn't help for some reason, so I tried changing the order in which the front and back objects are rendered and that seems to have fixed it. I'm not sure why that worked, though.

Sandbox: https://codesandbox.io/s/multiple-effect-passes-issue-forked-91t90?file=/src/App.js

**Reply by swingingtom** [reply]
Thanks a lot @vanruesc for the provided demo. I don't know if its only on my side, but in the demo you provided the first effect pass that glitch and depth of field is no more effective.

I will still try to change combinaisons order of passes, but I fear that my project will have to reduce its expected results. In the bright side, it might same some perfs.

Thanks !
